crypto/x509/x509_vfy.c

```diff
@@ -468,46 +468,59 @@ static int check_chain(X509_STORE_CTX *ctx)
         int ret;
 
         x = sk_X509_value(ctx->chain, i);
-        CB_FAIL_IF((ctx->param->flags & X509_V_FLAG_IGNORE_CRITICAL) == 0
-                       && (x->ex_flags & EXFLAG_CRITICAL) != 0,
-                   ctx, x, i, X509_V_ERR_UNHANDLED_CRITICAL_EXTENSION);
         CB_FAIL_IF(!allow_proxy_certs && (x->ex_flags & EXFLAG_PROXY),
                    ctx, x, i, X509_V_ERR_PROXY_CERTIFICATES_NOT_ALLOWED);
+
+        /*
+         * Skip further checks for directly trusted self-issued EE certs
+         * because RFC 5280 does not apply to them according RFC 6818 section 2.
+         */
+        if (num == 1 && (x->ex_flags & EXFLAG_CA) == 0
+                && (x->ex_flags & EXFLAG_SI) != 0)
+            continue;
+
+        /*
+         * Check the validity of the cert representation beforehand
+         * because otherwise the return code of X509_check_ca(x)
+         * and various other checks actually have no (useful) value
+         */
+        CB_FAIL_IF((x->ex_flags & EXFLAG_INVALID) != 0,
+                   ctx, x, i, X509_V_ERR_INVALID_CERTIFICATE_FORMAT);
+
         ret = X509_check_ca(x);
         switch (must_be_ca) {
-        case -1:
+        case -1: /*  we accept both CA and non-CA certificates */
             CB_FAIL_IF((ctx->param->flags & X509_V_FLAG_X509_STRICT) != 0
-                           && ret != 1 && ret != 0,
+                           && ret != X509_CA_TYPE_V3
+                           && ret != X509_CA_TYPE_NON_CA_OR_ERROR,
                        ctx, x, i, X509_V_ERR_INVALID_CA);
             break;
-        case 0:
-            CB_FAIL_IF(ret != 0, ctx, x, i, X509_V_ERR_INVALID_NON_CA);
+        case 0: /* we only accept non-CA certificates */
+            CB_FAIL_IF(ret != X509_CA_TYPE_NON_CA_OR_ERROR,
+                       ctx, x, i, X509_V_ERR_INVALID_NON_CA);
             break;
-        default:
+        default: /* we only accept CA certificates */
             /* X509_V_FLAG_X509_STRICT is implicit for intermediate CAs */
-            CB_FAIL_IF(ret == 0
+            CB_FAIL_IF(ret == X509_CA_TYPE_NON_CA_OR_ERROR
                        || ((i + 1 < num
                             || ctx->param->flags & X509_V_FLAG_X509_STRICT)
-                           && ret != 1), ctx, x, i, X509_V_ERR_INVALID_CA);
+                           && ret != X509_CA_TYPE_V3),
+                       ctx, x, i, X509_V_ERR_INVALID_CA);
             break;
         }
-        if (num > 1) {
-            /* Check for presence of explicit elliptic curve parameters */
-            ret = check_curve(x);
-            CB_FAIL_IF(ret < 0, ctx, x, i, X509_V_ERR_UNSPECIFIED);
-            CB_FAIL_IF(ret == 0, ctx, x, i, X509_V_ERR_EC_KEY_EXPLICIT_PARAMS);
-        }
```

Comment:
It seems to me that the original code here (added recently in 1.1.1, after I last took a deep dive into the X.509 verification code) may not have been doing what was intended.  The test `num > 1` tests the chain to see whether it has any CAs, and if so restricts *every* certificate in the chain (including the leaf) to not use explicit curve parameters.

Is this really the correct criterion? If we're also testing the leaf cert when `num > 1`, why not test it when `num == 1`? Alternatively, if the idea is to make sure none of the CA certs have been "tweaked", shouldn't the test be `i > 0` rather than `num > 1` and test everything but the leaf cert?

Finally, OpenSSL does not IIRC share the Windows `Curveball` bug, so it is not entirely clear why this has been added to OpenSSL.

The reason I mention all this, is that the new code is subtly different.  It no longer applies the `num > 1` constraint, but it is only enforced after self-issued EE certificates are exempted from further checks.  This means that with `partial_chain` and a non-self-issued, but trusted EE certificate will now also be restricted when it is the only one in the chain.

So there are many subtle edge conditions here, and I don't have a clear sense that anyone knows what the expected semantics are.  So I'd like to see that resolved.

Reply:
Good point that the intention which certs should actually be checked here under which circumstances is somewhat unclear.
AFAIR, I thought that the guard `num > 1` makes not much sense and thus simply removed it.
Yet to be on the conservative side, I just re-added it and added a TODO comment about it such that this PR is relieved from solving this pre-existing issue.